packages/types/lib/command.ts

```diff
@@ -0,0 +1,119 @@
+import {ConditionalPick, MultidimensionalReadonlyArray} from 'type-fest';
+import {Driver, DriverCommand} from './driver';
+import {Plugin, PluginCommand} from './plugin';
+
+/**
+ * Defines the shape of a payload for a {@linkcode MethodDef}.
+ */
+export interface PayloadParams {
+  wrap?: string;
+  unwrap?: string;
+  required?: ReadonlyArray<string> | MultidimensionalReadonlyArray<string, 2>;
+  optional?: ReadonlyArray<string> | MultidimensionalReadonlyArray<string, 2>;
+  validate?: (obj: any, protocol: string) => boolean | string | undefined;
+  makeArgs?: (obj: any) => any;
+}
+/**
+ * A mapping of URL paths to HTTP methods to either a {@linkcode DriverMethodDef} or {@linkcode PluginMethodDef}.
+ */
+export type MethodMap<T extends Plugin | Driver> = T extends Plugin
+  ? Readonly<PluginMethodMap<T>>
+  : T extends Driver
+  ? Readonly<DriverMethodMap<T>>
+  : never;
+
+/**
+ * A {@linkcode MethodMap} for a {@linkcode Driver}.
+ */
+export interface DriverMethodMap<T extends Driver> {
+  [key: string]: {
+    GET?: DriverMethodDef<T>;
+    POST?: DriverMethodDef<T>;
+    DELETE?: DriverMethodDef<T>;
+  };
+}
+
+/**
+ * Both {@linkcode DriverMethodDef} and {@linkcode PluginMethodDef} share these properties.
+ */
+export interface BaseMethodDef {
+  /**
+   * If true, this `Method` will never proxy.
+   */
+  readonly neverProxy?: boolean;
+  /**
+   * Specifies shape of payload
+   */
+  readonly payloadParams?: PayloadParams;
+}
+
+/**
+ * A definition of an exposed API command in a {@linkcode Driver}.
+ */
+export interface DriverMethodDef<T extends Driver> extends BaseMethodDef {
+  /**
+   * Name of the command.
+   */
+  readonly command?: keyof ConditionalPick<Required<T>, DriverCommand>;
+}
+
+/**
+ * A definition of an exposed API command in a {@linkcode Plugin}.
+ */
+export interface PluginMethodDef<T extends Plugin> extends BaseMethodDef {
+  /**
+   * Name of the command.
+   */
+  readonly command?: keyof ConditionalPick<Required<T>, DriverCommand>;
+}
+
+/**
+ * A mapping of URL paths to HTTP methods to {@linkcode PluginMethodDef}.
+ */
+export interface PluginMethodMap<T extends Plugin> {
+  [key: string]: {
+    GET?: PluginMethodDef<T>;
+    POST?: PluginMethodDef<T>;
+    DELETE?: PluginMethodDef<T>;
+  };
+}
+
+export interface ExecuteMethodDef<Ext extends Plugin | Driver> {
+  command: keyof ConditionalPick<
+    Required<Ext>,
+    Ext extends Plugin ? PluginCommand : Ext extends Driver ? DriverCommand : never
+  >;
+}
+
+/**
+ * Properties shared by execute method definitions in both plugins and drivers
+ */
+export interface BaseExecuteMethodDef {
+  params?: {
+    required?: ReadonlyArray<string>;
+    optional?: ReadonlyArray<string>;
+  };
+}
+
+/**
+ * A definition of an execute method in a {@linkcode Driver}.
+ */
+export interface DriverExecuteMethodDef<T extends Driver> extends BaseExecuteMethodDef {
+  command: keyof ConditionalPick<T, DriverCommand>;
```

Comment:
fwiw this checks the 'command' prop of the method map and ensures the name you put in there is a method which  satisfies the `DriverCommand` type (basically just that it's `async`).  this is more useful for `PluginCommand` since there are implicit parameters on each.